File: BackendDotNet/Controllers/DonorController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using BackendDotNet.DTOs.Request;
using BackendDotNet.DTOs.Response;
using BackendDotNet.Models;
using BackendDotNet.Services;

namespace BackendDotNet.Controllers
{
    [ApiController]
    [Route("api/donor")]
    [Authorize(Roles = "ROLE_DONOR")]
    public class DonorController : ControllerBase
    {
        private readonly IDonorService _donorService;

        public DonorController(IDonorService donorService)
        {
            _donorService = donorService;
        }

        [HttpGet("profile")]
        public async Task<ActionResult<DonorProfile>> GetMyProfile()
        {
            try
            {
                var profile = await _donorService.GetMyProfileAsync();
                return Ok(profile);
            }
            catch (InvalidOperationException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving profile", error = ex.Message });
            }
        }

        [HttpPost("appointment")]
        public async Task<ActionResult<DonationAppointment>> BookAppointment([FromBody] AppointmentCreateRequest request)
        {
            try
            {
                var appointment = await _donorService.BookAppointmentAsync(request);
                return StatusCode(201, appointment);
            }
            catch (InvalidOperationException ex)
            {
                return BadRequest(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error booking appointment", error = ex.Message });
            }
        }

        [HttpGet("donations/history")]
        public async Task<ActionResult<List<DonationRecordDto>>> GetMyDonationHistory()
        {
            try
            {
                var history = await _donorService.GetMyDonationHistoryAsync();
                return Ok(history);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving donation history", error = ex.Message });
            }
        }

        [HttpPatch("appointment/{appointmentId}/cancel")]
        public async Task<ActionResult<DonationAppointment>> CancelAppointment(long appointmentId)
        {
            try
            {
                var cancelledAppointment = await _donorService.CancelAppointmentAsync(appointmentId);
                return Ok(cancelledAppointment);
            }
            catch (InvalidOperationException ex)
            {
                return BadRequest(new { message = ex.Message });
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error cancelling appointment", error = ex.Message });
            }
        }

        [HttpGet("appointments/scheduled")]
        public async Task<ActionResult<List<DonationAppointment>>> GetMyScheduledAppointments()
        {
            try
            {
                var appointments = await _donorService.GetMyScheduledAppointmentsAsync();
                return Ok(appointments);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving scheduled appointments", error = ex.Message });
            }
        }
    }
}

File: backend spring/src/main/java/com/bloodbank/controller/DonorController.java
// File: src/main/java/com/bloodbank/controller/DonorController.java

package com.bloodbank.controller;

import com.bloodbank.dto.request.AppointmentCreateRequest;
import com.bloodbank.dto.response.DonationRecordDto;
import com.bloodbank.entity.DonationAppointment;
import com.bloodbank.entity.DonorProfile;
import com.bloodbank.service.DonorService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/donor")
@PreAuthorize("hasRole('DONOR')")
public class DonorController {

    @Autowired
    private DonorService donorService;

    // Existing methods
    @GetMapping("/profile")
    public ResponseEntity<DonorProfile> getMyProfile() {
        DonorProfile donorProfile = donorService.getMyProfile();
        return ResponseEntity.ok(donorProfile);
    }

    @PostMapping("/appointment")
    public ResponseEntity<DonationAppointment> bookAppointment(@RequestBody AppointmentCreateRequest request) {
        DonationAppointment newAppointment = donorService.bookAppointment(request);
        return new ResponseEntity<>(newAppointment, HttpStatus.CREATED);
    }

    @GetMapping("/donations/history")
    public ResponseEntity<List<DonationRecordDto>> getMyDonationHistory() {
        return ResponseEntity.ok(donorService.getMyDonationHistory());
    }

    // --- NEW ENDPOINTS ---

    @PatchMapping("/appointment/{appointmentId}/cancel")
    public ResponseEntity<DonationAppointment> cancelAppointment(@PathVariable Long appointmentId) {
        DonationAppointment cancelledAppointment = donorService.cancelAppointment(appointmentId);
        return ResponseEntity.ok(cancelledAppointment);
    }

    @GetMapping("/appointments/scheduled")
    public ResponseEntity<List<DonationAppointment>> getMyScheduledAppointments() {
        return ResponseEntity.ok(donorService.getMyScheduledAppointments());
    }
}